Update hyper to v0.12

arsing opened the pull request:
The main differences are:

- hyper requires service futures to be Send, which rippled through most types
  in the codebase. This is because it uses the new tokio thread-pool runtime
  under the hood instead of the old tokio-core single-threaded executor.

- Replace our use of tokio-core with tokio, and update other tokio-related
  crates, for the same reason. Tests continue to use the single-threaded
  runtime.

- The implicit tokio runtime means that the code no longer needs to pass
  `tokio_core::Handle` around, so it has been removed from those structs and
  constructors.

- hyper's `Connect` and `Service` traits now have a different meaning.

  - `Connect` is only meant to be used with things that provide a transport,
    and not for clients in general.

  - `Service` is only meant for server-type services, must be Send, and takes
    `&mut self` since new Service values are generated by the NewService impl
    for each client connection.

  The code relied on a default `Connect` impl for all `Service` impls that
  took a `Uri` as input. The equivalent of this for most cases is now provided
  through our own new `ClientImpl` trait.

- hyper no longer provides HTTP header types. Use the typed-headers crate
  for these. However `typed-headers::Authorization` is not as flexible
  as hyper 0.11's type was (does not allow arbitrary values like our
  SAS tokens), so this has been reverted to use raw headers.

- ~~The hyper-proxy crate on crates.io currently does not support
  hyper 0.12 - there is a PR in progress for it. For now I've reimplemented
  it as a new crate using code based on reqwest.~~

---

TODOs:

- ~~Consolidate use of `tokio` with specialized crates like `tokio-io`. `tokio` mostly re-exports all of them so either just use `tokio` everywhere or just the specialized crates everywhere (except for tests and `iotedged`).~~

- ~~clippy~~

- ~~rustfmt~~

- ~~Figure out why default parallel codegen crashes the linker in the Linux test job.~~

- ~~Update hyper-proxy to v0.5.0 that should support hyper 0.12~~

- ~~Either wait for https://github.com/tokio-rs/tokio/pull/672 , or disable keep_alive on both `hyper::Client` that use `UrlConnector`~~

arsing commented:
Disabling parallel codegen and incremental builds seems to have fixed the crashof ld in the Linux test job. The abrupt crash might have been because the VSTS VM was running out of memory or something. (Disabling parallel codegen and incremental builds reduces the number of objects the linker needs to link).

I'll leave it as-is for now but I assume we don't want to merge it like this. Maybe we can make it only get set for VSTS builds...

arsing commented on edgelet/Cargo.toml:
```diff
@@ -23,3 +23,11 @@ members = [
+
+[profile.dev]
+codegen-units = 1
+incremental = false
```
This is part of the disable-parallel-codegen commit. I'm going to move this to the CI script.

myagley replied:
Ok cool

arsing commented on edgelet/docker-rs/src/apis/client.rs:
```diff
@@ -1,30 +1,30 @@
+impl<C: hyper::client::connect::Connect + 'static> APIClient<C> {
     pub fn new(configuration: Configuration<C>) -> APIClient<C> {
-        let rc = Rc::new(configuration);
+        let configuration = Arc::new(configuration);
```
The `Send` requirement in `hyper`, but yes. All the `Rc` -> `Arc` are because of that.

arsing commented on edgelet/docker-rs/src/apis/container_api.rs:
```diff
@@ -196,11 +200,13 @@ impl<C: hyper::client::Connect> ContainerApi for ContainerApiClient<C> {
-        let mut req = hyper::Request::new(method, uri.unwrap());
+        let mut req = hyper::Request::new(hyper::Body::empty());
+        *req.method_mut() = method;
+        *req.uri_mut() = uri.unwrap();
```
It does, but it doesn't support typed headers (since `typed-headers` only adds functions to the `HeaderMap`, not the `Builder`, so I'll still have to mutate `headers_mut()` after building. Will do.

arsing commented on edgelet/dps/src/registration.rs:
```diff
@@ -442,52 +442,54 @@ fn get_device_info(
-                    )
+                    let mut response =
+                        Response::new(serde_json::to_string(&result).unwrap().into());
+                    *response.status_mut() = StatusCode::UNAUTHORIZED;
```
I think I did this because the Response builder enforces having to deal with a Result - `Builder::build()` returns a Result in case you configured it wrong. Will double-check.

arsing replied:
It's a test using `unwrap()` anyway, so I changed it to use the builder.

arsing commented on edgelet/edgelet-http-mgmt/src/server/identity/create.rs:
```diff
@@ -36,27 +35,27 @@ where
             read_request(req)
                 .and_then(move |spec| {
-                    let mut rid = id_mgr.borrow_mut();
+                    let mut rid = id_mgr.lock().unwrap();
```
Yesbut I should change it to an `expect()` with that message

arsing commented on edgelet/edgelet-http-workload/src/error.rs:
```diff
@@ -145,12 +145,6 @@ impl From<UtilsError> for Error {
     }
 }
 
-impl From<Error> for HyperError {
```
Well, it's just impossible now since `hyper::Error` is opaque.

myagley replied:
Oh. Gotcha

arsing commented on edgelet/edgelet-http/src/client.rs:
```diff
@@ -20,67 +18,96 @@ pub trait TokenSource {
 }
 
-pub struct Client<S, T>
+pub trait ClientImpl: Send + Sync {
```
See the PR description.

myagley replied:
yes, I should have read it more closely. :)

arsing replied:
The problem is the `&self` -> `&mut self`. A bunch of places relied on using `Arc<TypeThatImplsService>` and I didn't want to have to make them `Clone` / wrap `NewService` instead of `Service` (the latter especially since `NewService::new_service` is async).

arsing commented on edgelet/edgelet-http/src/util/connector.rs:
```diff
@@ -83,33 +85,49 @@ impl UrlConnector {
-                    .call(uri)
-                    .and_then(|tcp_stream| Ok(StreamSelector::Tcp(tcp_stream))),
-            ) as Self::Future,
+    fn connect(&self, dst: Destination) -> Self::Future {
```
Can't borrow `dst` via `dst.scheme()` for the whole `match` and move it to `FooConnector.connect()` inside one of its arms. https://play.rust-lang.org/?version=nightly&mode=debug&edition=2015 (NLL allows it.)

myagley commented on edgelet/docker-rs/src/apis/container_api.rs:
```diff
@@ -611,11 +627,13 @@ impl<C: hyper::client::Connect> ContainerApi for ContainerApiClient<C> {
         if let Some(ref user_agent) = configuration.user_agent {
             req.headers_mut()
-                .set(UserAgent::new(Cow::Owned(user_agent.clone())));
+                .append(http::header::USER_AGENT, user_agent.parse().unwrap());
```
Or I guess can we change all of the `unwrap`s to `expect` so that we can annotate why we think they are safe.

arsing commented on edgelet/edgelet-http-workload/src/server/mod.rs:
```diff
@@ -30,11 +31,17 @@ pub struct WorkloadService {
+        key_store: &K,
+        hsm: H,
+        runtime: &M,
+    ) -> impl Future<Item = Self, Error = failure::Error>
```
hyper's `NewService::new_service` is now async.

arsing commented on edgelet/edgelet-http/src/client.rs:
```diff
@@ -125,48 +154,50 @@ where
                 // add user agent header
                 if let Some(ref user_agent) = self.user_agent {
-                    req.headers_mut().set(UserAgent::new(user_agent.clone()));
+                    req.headers_mut().append(http::header::USER_AGENT, user_agent.parse().unwrap());
```
The builder collects all header errors into the final `.body().expect()`, so all the user agent `.parse().unwrap()` are gone.

arsing commented on edgelet/edgelet-http/src/lib.rs:
```diff
@@ -106,59 +107,73 @@ impl IntoResponse for Response<Body> {
+                            }
+                        })
+                });
+            tokio::spawn(fut);
```
It's the equivalent of `tokio_core::Handle::spawn()` if that's what you mean. It looks up the global runtime and spawns on it.

arsing commented on edgelet/edgelet-http/src/route/mod.rs:
```diff
@@ -132,23 +143,23 @@ impl<R> Service for RouterService<R>
-                        .body(Body::default())
-                        .map_err(|_| HyperError::Status),
+                        .body(Body::empty())
+                        .unwrap(),
```
Yes, and there's no other builder calls that could have been given invalid values (like `.header()` with an invalid header). Changed to an `.expect()` that documents this.

arsing commented on edgelet/edgelet-http-mgmt/src/server/mod.rs:
```diff
@@ -4,16 +4,19 @@ mod identity;
+use failure;
+use futures::{future, Future};
+use hyper::{
+    service::{NewService, Service},
```
Will do.

I was going to add a rustfmt.toml with `merge_imports = false` to future-proof it, but turns out that option requires nightly atm.

arsing commented on edgelet/edgelet-http-mgmt/src/server/mod.rs:
```diff
@@ -64,30 +69,34 @@ impl ManagementService {
+        router
+            .new_service()
+            .map(|inner| ManagementService { inner })
+            .map_err(failure::Error::from_boxed_compat)
```
`<Router as NewService>::InitError is Box<StdError>`. `failure::Error::from_boxed_compat` converts `Box<StdError>` to `failure::Error`

arsing commented on edgelet/iotedged/src/lib.rs:
```diff
@@ -489,23 +493,36 @@ where
+                Err(())
+            }
+        });
+    tokio_runtime.block_on(shutdown.select(services).then(|result| match result {
```
This `select` is wrong, since it means the services don't get to run their shutdown at all. In particular the `ModuleRuntime::stop` does not get invoked so the container is left in running state.

arsing commented:
Squashed and rebased, @myagley Can you approve again please?

@darobs Can you verify the change in https://github.com/Azure/iotedge/commit/762b8d917b0708c2fb99c7f7b7af492e82035eeb

arsing commented on edgelet/docker-rs/src/apis/client.rs:
```diff
@@ -1,30 +1,30 @@
-pub struct APIClient<C: hyper::client::Connect> {
-    configuration: Rc<Configuration<C>>,
+pub struct APIClient<C: hyper::client::connect::Connect> {
+    configuration: Arc<Configuration<C>>,
```
Yes.
